core/debug/src/debug_box.c

```diff
@@ -39,6 +39,51 @@ void debug_reboot(TResetReason reason)
 
 uint32_t g_debug_interrupt_sp[UVISOR_MAX_BOXES];
 
+void debug_print(const uint8_t * message_buffer, uint32_t size)
+{
+	/* Abort printing if debug box wasn't initialized yet. */
+	if (!g_debug_box.initialized) {
+		return;
+	}
+
+	/* Debug print relies on "deprivilege and return" flow that will currently
+	 * only work if:
+	 * - executed with exceptions enabled
+     * - executed from within the exception context, i.e. when IPSR isn't 0
+     * - the current exception priority is lower than of Hard Fault
+	 * These limitation are dictated by the fact that SVC is used for
+	 * deprivileging and in order for it to work SVC must be executed with
+	 * exceptions enabled and its priority must be higher than of the current exception.
+	 * SVC's priority is set to be higher than of any other exception with
+	 * programmable priority but this still can't beat NMI and Hard Fault.
+	 */
+
+	/* Make sure we're within an exception and exception are enabled. */
+	int exception = (int)(__get_IPSR() & 0x1FF);
+	if (!exception || __get_PRIMASK() || __get_FAULTMASK())	{
+		return;
+	}
+
+	/* Bring exception number to CMSIS format and check exception's priority. */
+	exception -= NVIC_OFFSET;
+	if (exception == NonMaskableInt_IRQn || exception == HardFault_IRQn) {
+		return;
+	}
+
+	/* Place the message on the interrupts stack allocating
+	 * the size rounded up to the next value divisible by 4. */
+	uint32_t sp = g_debug_interrupt_sp[g_debug_box.box_id];
+	g_debug_interrupt_sp[g_debug_box.box_id] -= ((size + 3) / 4) * 4;
+	void *message = (void *)g_debug_interrupt_sp[g_debug_box.box_id];
```

Comment:
Not sure where exactly you propose to use sp.
If in line
void *message = (void *)g_debug_interrupt_sp[g_debug_box.box_id];
then sp without manipulations won't be good since we just updated g_debug_interrupt_sp[g_debug_box.box_id] and sp contains the original value.

Reply:
Couldn't you assign `message` before you modify `g_debug_interrupt_sp[g_debug_box.box_id]`?

Reply:
Maybe I wasn't clear enough, but we do need to assign modified value of g_debug_interrupt_sp[g_debug_box.box_id] into message (the new stack top after reserving space for the message).
Still can't see a way to make the code prettier.